### PR TITLE
Add date/time parsing and formatting functions

### LINKED ISSUE #166: Configurable date formats and time zones
The Go implementation of JSONata is extending the date functions to allow user-defined date formats and time zones. We think this would be a useful addition to the JSONata standard so I'm opening this issue as a request for comment.

The relevant functions are:

    $now()  // returns the current time as a string

    $fromMillis(number)  // converts a Unix time to a string
    
    $toMillis(timestamp)  // converts a string to a Unix time

All of these functions use ISO 8601 date format and UTC. We plan to add the following optional arguments:

    $now([picture [, timezone]])
    
    $fromMillis(number [, picture [, timezone]])
    
    $toMillis(timestamp [, picture])

The `picture` argument specifies a date format using a subset of [XPath's date/time syntax](https://www.w3.org/TR/xpath-functions-31/#date-time-examples). The `timezone` argument specifies the time zone to use when outputting a date as a string. It should handle IANA identifiers like "America/Chicago", time zone abbreviations like "EST", and explicit time offsets like "-0500".

To ensure backwards compatibility, the new arguments will be optional with defaults that make existing expressions generate the same results as before.

### PR DESCRIPTION
**Pull request by andrew-coleman**
Extend the `$toMillis` and`$fromMillis` functions to support parsing and formatting of timestamps
as defined by a 'picture string' parameter.  The specification for this is aligned with XPath's [fn:format-dateTime](https://www.w3.org/TR/xpath-functions-31/#func-format-dateTime) function.

Also added new `$formatInteger` and `$parseInteger` functions as specified by XPath's [fn:format-Integer](https://www.w3.org/TR/xpath-functions-31/#func-format-integer) function.  This implementation of these were required for implementing the date/time functions, so it makes sense to expose them independently.

The format of the test suite was extended to support test case files that contain arrays of test definitions.

The code that implements these new functions is in a separate file (datetime.js).  This is the first time I have split the core jsonata implementation out from the single jsonata.js file.  The build scripts have been updated to handle this.

resolves #166

### CODE REVIEW

File: src/jsonata.js [hunk 1]
```diff
@@ -5098,9 +5116,9 @@ var jsonata = (function() {
         var environment = createFrame(staticFrame);
 
         var timestamp = new Date(); // will be overridden on each call to evalute()
-        environment.bind('now', defineFunction(function() {
-            return timestamp.toJSON();
-        }, '<:s>'));
+        environment.bind('now', defineFunction(function(picture, timezone) {
+            return typeof picture === 'undefined' ? timestamp.toJSON() : datetime.formatDateTime(timestamp.getTime(), picture, timezone);
```
Comment (mattbaileyuk) [inline src/jsonata.js]: Same here for $now()

File: src/jsonata.js [hunk 2]
```diff
@@ -4832,38 +4829,49 @@ var jsonata = (function() {
...
      */
-    function functionFromMillis(millis) {
+    function functionFromMillis(millis, picture, timezone) {
         // undefined inputs always return undefined
         if(typeof millis === 'undefined') {
             return undefined;
         }
 
-        return new Date(millis).toISOString();
+        if(typeof picture === 'undefined') {
```
[the remark this answers is not in the document]
Comment (andrew-coleman) [reply src/jsonata.js]: good point.  undefined picture now defaults to ISO 8601 for both of these functions

File: .travis.yml [hunk 3]
```diff
@@ -1,9 +1,9 @@
 language: node_js
 node_js:
+- '11'
 - '10'
 - '8'
 - '6'
-- '4'
```
[the remark this answers is not in the document]
Comment (andrew-coleman) [reply .travis.yml]: For build/test yes.  Runtime should continue to work.  But node 4 itself is out of LTS.
```
  "engines": {
    "node": ">= 4"
  }
```

[the remark this answers is not in the document]
Comment (mattbaileyuk) [reply .travis.yml]: Node 4 is well out of LTS, and node 6 is the "maintenance LTS"  until it is dropped in April, so completely in support of dropping 4. However, if we don't drop it completely then there's a risk of introducing something which is not compatible with node 4; this happened in [this recent PR](https://github.com/jsonata-js/jsonata/pull/263) where one of the commits didn't work on that version.

[the remark this answers is not in the document]
Comment (andrew-coleman) [reply .travis.yml]: Fair point. My main concern was not breaking Node-RED.  But it looks like they are moving to node 8 for their next release.

**Comment by mattbaileyuk** [comment]
Just need some more doc updates to follow this 😄

**Pull request merged**